File: src/types/delivery.ts
export type DeliveryPriority = 'low' | 'medium' | 'high' | 'urgent';

export type DeliveryStatus = 'pending' | 'assigned' | 'in_progress' | 'delivered' | 'cancelled';

export interface DeliveryCoordinates {
	latitude: number | null;
	longitude: number | null;
}

export interface DeliveryVehicle {
	id: string;
	registration?: string | null;
}

export interface DeliveryDriver {
	id: string;
	fullName?: string | null;
}

export interface Delivery {
	id: string;
	reference: string;
	customer: string;
	pickupAddress: string;
	dropoffAddress: string;
	weightKg: number | string;
	priority: DeliveryPriority;
	status: DeliveryStatus;
	vehicleId?: string | null;
	driverId?: string | null;
	vehicle?: DeliveryVehicle | null;
	driver?: DeliveryDriver | null;
	scheduledAt: string;
	pickupLatitude: number | null;
	pickupLongitude: number | null;
	dropoffLatitude: number | null;
	dropoffLongitude: number | null;
	createdAt?: string;
	updatedAt?: string;
}

export interface DeliveryPayload {
	customer: string;
	pickupAddress: string;
	dropoffAddress: string;
	pickupLatitude: number | null;
	pickupLongitude: number | null;
	dropoffLatitude: number | null;
	dropoffLongitude: number | null;
	weightKg: number | string;
	priority: DeliveryPriority;
	scheduledAt: string;
}

export interface DeliveryAssignmentPayload {
	vehicleId: string;
	driverId: string;
}

File: app/models/delivery.py
from datetime import datetime
from enum import Enum


class DeliveryStatus(str, Enum):
    pending = "pending"
    assigned = "assigned"
    picked = "picked"
    in_progress = "in_progress"
    delivered = "delivered"
    cancelled = "cancelled"


class Delivery:
    def __init__(
        self,
        reference: str,
        customer: str,
        pickup_address: str,
        pickup_address_lat: float,
        pickup_address_lng: float,
        dropoff_address: str,
        dropoff_address_lat: float,
        dropoff_address_lng: float,
        weight_kg: float,
        priority: str,
        scheduled_at: datetime,
        status: DeliveryStatus = DeliveryStatus.pending,
        vehicle_id: str | None = None,
        driver_id: str | None = None,
        delivered_at: datetime | None = None,
        created_at: datetime | None = None,
        updated_at: datetime | None = None,
        id: str | None = None,
    ):
        self.id = id
        self.reference = reference
        self.customer = customer
        self.pickup_address = pickup_address
        self.pickup_address_lat = pickup_address_lat
        self.pickup_address_lng = pickup_address_lng
        self.dropoff_address = dropoff_address
        self.dropoff_address_lat = dropoff_address_lat
        self.dropoff_address_lng = dropoff_address_lng
        self.weight_kg = weight_kg
        self.priority = priority
        self.scheduled_at = scheduled_at
        self.status = status
        self.vehicle_id = vehicle_id
        self.driver_id = driver_id
        self.delivered_at = delivered_at
        self.created_at = created_at or datetime.utcnow()
        self.updated_at = updated_at or datetime.utcnow()
